cloud-controller-manager/do/firewall_controller.go

```diff
@@ -0,0 +1,364 @@
+		if err != nil {
+			klog.Errorf("failed to run firewall controller loop: %v", err)
+		}
+	}, fwReconcileFrequency, stopCh)
+}
+
+func (fc *FirewallController) actualRun(stopCh <-chan struct{}, fm *firewallManagerOp, fwReconcileFrequency time.Duration) error {
+	ctx, cancel := context.WithTimeout(context.Background(), fwReconcileFrequency)
+	defer cancel()
+
+	currentFirewall, err := fc.fwManager.Get(ctx)
+	if err != nil {
+		return fmt.Errorf("failed to get worker firewall: %s", err)
+	}
+	if currentFirewall != nil {
+		if fm.fwCache.isEqual(currentFirewall) {
+			return nil
+		}
+	}
+	fm.fwCache.updateCache(currentFirewall)
+	err = fc.ensureReconciledFirewall(ctx)
+	if err != nil {
+		return fmt.Errorf("failed to reconcile worker firewall: %s", err)
+	}
+
+	klog.Info("successfully reconciled firewall")
+	return nil
+}
+
+// Get returns the current public access firewall representation.
+func (fm *firewallManagerOp) Get(ctx context.Context) (*godo.Firewall, error) {
+	// check cache and query the API firewall service to get firewall ID, if it exists. Return it. If not, continue.
+	fw := fm.fwCache.getCachedFirewall()
+	if fw != nil {
+		fw, resp, err := fm.client.Firewalls.Get(ctx, fw.ID)
+		if err != nil && (resp == nil || resp.StatusCode != http.StatusNotFound) {
+			return nil, fmt.Errorf("could not get firewall: %v", err)
+		}
+		if resp.StatusCode == http.StatusNotFound {
+			klog.Warningf("unable to retrieve firewall by ID because it no longer exists")
+		}
+		if fw != nil {
+			return fw, nil
+		}
+	}
+
+	// iterate through firewall API provided list and return the firewall with the matching firewall name.
+	f := func(fw godo.Firewall) bool {
+		return fw.Name == fm.workerFirewallName
+	}
+	klog.Info("filtering firewall list for the firewall that has the expected firewall name")
+	fw, err := filterFirewallList(ctx, fm.client, f)
+	if err != nil {
+		return nil, fmt.Errorf("failed to retrieve list of firewalls from DO API: %v", err)
+	}
+	return fw, nil
+}
+
+// Set applies the given inbound rules to the public access firewall when the current rules and target rules differ.
+// Set will also reconcile away any changes to the outbound rules, name and tags.
+func (fm *firewallManagerOp) Set(ctx context.Context, fr *godo.FirewallRequest) error {
+	targetFirewall := fm.fwCache.getCachedFirewall()
+
+	if targetFirewall != nil {
+		equal, _ := isEqual(targetFirewall, fr)
+		if equal {
+			// A locally cached firewall with matching rules, correct name and tags means there is nothing to update.
+			return nil
+		}
+		// A locally cached firewall exists, but the inbound rules don't match the expected
+		// service inbound rules, outbound rules, name or tags. So we need to use the locally
+		// cached firewall ID to attempt to update the firewall APIs representation of the
+		// firewall with the new rules.
+		currentFirewall, resp, err := fm.client.Firewalls.Update(ctx, targetFirewall.ID, fr)
+		if err != nil {
```

Comment:
Let's include a `if err == nil { ... }`  block here where we just log the successful update operation.